### PR TITLE
Update MLPerfCallback for v2.1

### PR DESCRIPTION
**Pull request by hanlint**
# What does this PR do?

Updates the MLPerfCallback by adding `bert` support, and updating the MLperf compliance checker to the v2.1 ruleset. Also fixes the tests to be passing. 

Detailed fixes to `MLPerfCallback`:
* `train_samples` is now logged properly when using `IterableDataset`, which has a per-device length.
* the BERT benchmark logs time in units of samples, whereas resnet logs in epochs
* uploading the results files now occurs at the end of every `eval` step, and also when the callback is closed.
* adds option `exit_at_target` to exit training when target is met. Default: False

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Review by dskhudia — APPROVED** [review]
LGTM! Thank you

**Review by dakinggg — APPROVED** [review]
LGTM, one comment because I just merged #1601

**Review by eracah — APPROVED** [review]
LGTM

File: composer/callbacks/mlperf.py [hunk 1]
```diff
@@ -320,8 +320,8 @@ def epoch_start(self, state: State, logger: Logger) -> None:
 
     def epoch_end(self, state: State, logger: Logger) -> None:
         if _global_rank_zero():
-            self.mllogger.event(key=constants.EPOCH_STOP, metadata={'epoch_num': self._get_time(state)})
-            logger.file_artifact(artifact_name=self.upload_name, file_path=self.filename)
+            self.mllogger.event(key=constants.EPOCH_STOP, metadata={'epoch_num': state.timestamp.epoch.value})
```
Comment (dakinggg) [inline composer/callbacks/mlperf.py:323]: I think you probably didn't mean to change this back to `state.timestamp.epoch.value` in the merge?

File: composer/callbacks/mlperf.py [hunk 2]
```diff
@@ -297,37 +311,44 @@ def fit_start(self, state: State, logger: Logger) -> None:
...
+                                metadata={'epoch_num': self._get_time(state)})
+            self.mllogger.event(key=constants.BLOCK_STOP, metadata={'first_epoch_num': self._get_time(state)})
 
             if accuracy > self.target and not self.success:
                 self.mllogger.event(key=constants.RUN_STOP, metadata={'status': 'success'})
                 self.mllogger.logger.removeHandler(self._file_handler)
                 self.success = True  # only log once
 
+            # upload to object store after eval complete
+            logger.file_artifact(artifact_name=self.upload_name, file_path=self.filename)
```
[the remark this answers is not in the document]
Comment (dakinggg) [reply composer/callbacks/mlperf.py:346]: I believe this still needs to be refactored to `logger.upload_file(remote_file_name=self.upload_name, file_path=self.filename)`.

File: composer/callbacks/mlperf.py (hunk 1 above)
Comment (hanlint) [reply composer/callbacks/mlperf.py]: ah, good catch, thank you!

**Pull request merged**